persistence/sql/persister_device.go

```diff
@@ -0,0 +1,305 @@
+	r := DeviceRequestSQL{}
+	err := p.QueryWithNetwork(ctx).Where("device_code_signature = ?", signature).First(&r)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, errorsx.WithStack(fosite.ErrNotFound)
+	}
+	if err != nil {
+		return nil, sqlcon.HandleError(err)
+	}
+	if !r.DeviceCodeActive {
+		fr, err := r.toRequest(ctx, session, p)
+		if err != nil {
+			return nil, err
+		}
+		return fr, errorsx.WithStack(fosite.ErrInactiveToken)
+	}
+
+	return r.toRequest(ctx, session, p)
+}
+
+func (p *Persister) findDeviceAuthSessionByUserCodeSignature(ctx context.Context, signature string, session fosite.Session) (fosite.DeviceRequester, error) {
+	r := DeviceRequestSQL{}
+	err := p.QueryWithNetwork(ctx).Where("user_code_signature = ?", signature).First(&r)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, errorsx.WithStack(fosite.ErrNotFound)
+	}
+	if err != nil {
+		return nil, sqlcon.HandleError(err)
+	}
+
+	fr, err := r.toRequest(ctx, session, p)
+	if err != nil {
+		return nil, err
+	}
+	if r.UserCodeState != fosite.UserCodeUnused {
+		return fr, errorsx.WithStack(fosite.ErrInactiveToken)
+	}
+
+	return fr, err
+}
+
+func (p *Persister) findDeviceAuthSessionByRequestID(ctx context.Context, requestID string, session fosite.Session) (fosite.DeviceRequester, string, error) {
+	r := DeviceRequestSQL{}
+	err := p.QueryWithNetwork(ctx).Where("request_id = ?", requestID).First(&r)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, "", errorsx.WithStack(fosite.ErrNotFound)
+	}
+	if err != nil {
+		return nil, "", sqlcon.HandleError(err)
+	}
+	if !r.DeviceCodeActive {
+		fr, err := r.toRequest(ctx, session, p)
+		if err != nil {
+			return nil, "", err
+		}
+		return fr, r.ID, errorsx.WithStack(fosite.ErrInactiveToken)
+	}
+
+	fr, err := r.toRequest(ctx, session, p)
+	if err != nil {
+		return nil, "", err
+	}
+	return fr, r.ID, nil
+}
+
+// CreateDeviceCodeSession creates a new device code session and stores it in the database
+func (p *Persister) CreateDeviceAuthSession(ctx context.Context, deviceCodeSignature, userCodeSignature string, requester fosite.DeviceRequester) (err error) {
+	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.CreateDeviceCodeSession")
+	defer otelx.End(span, &err)
+	return p.createDeviceAuthSession(ctx, deviceCodeSignature, userCodeSignature, requester, sqlTableDeviceAuthCodes, requester.GetSession().GetExpiresAt(fosite.DeviceCode).UTC())
+}
+
+// UpdateDeviceCodeSessionBySignature updates a device code session by the device_code signature
+func (p *Persister) UpdateDeviceCodeSessionBySignature(ctx context.Context, signature string, requester fosite.DeviceRequester) (err error) {
+	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.UpdateDeviceCodeSessionBySignature")
+	defer otelx.End(span, &err)
+
+	req, err := p.sqlDeviceSchemaFromRequest(ctx, signature, "", requester, requester.GetSession().GetExpiresAt(fosite.DeviceCode).UTC())
+	if err != nil {
+		return err
+	}
+
+	stmt := fmt.Sprintf(
```

Comment:
Not sure I understand, afaict all update statements are generated this way. Or do you mean to exec the query in an one-liner to avoid the creation of the `stmt` var?